src/Raven.Server/Documents/Sharding/Handlers/ShardedIncomingReplicationHandler.cs

```diff
@@ -73,21 +73,69 @@ private TcpConnectionInfo GetConnectionInfo(ShardReplicationNode node)
 
         protected override void EnsureNotDeleted(string nodeTag) => _parent.EnsureNotDeleted(_parent.Server.NodeTag);
 
-        protected override void InvokeOnFailed(Exception exception)
-        {
-            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Handle replication failures");
-        }
+        protected override void InvokeOnFailed(Exception exception) => _parent.InvokeOnFailed(this, exception);
 
         protected override void HandleHeartbeatMessage(TransactionOperationContext jsonOperationContext, BlittableJsonReaderObject blittableJsonReaderObject)
         {
-            DevelopmentHelper.ShardingToDo(DevelopmentHelper.TeamMember.Shiran, DevelopmentHelper.Severity.Normal, "Find a way to implement change vector updates for tackling optimization issues");
+            blittableJsonReaderObject.TryGet(nameof(ReplicationMessageHeader.DatabaseChangeVector), out string changeVector);
+
+            var batches = new ReplicationBatch[_parent.Context.ShardCount];
+            var tasks = new Task[batches.Length];
+            for (int i = 0; i < batches.Length; i++)
+            {
+                batches[i] = new ReplicationBatch
+                {
+                    LastAcceptedChangeVector = changeVector ?? _lastAcceptedChangeVector,
+                    LastSentEtagFromSource = _lastAcceptedEtag
+                };
+                tasks[i] = _handlers[i].SendBatch(batches[i]);
+            }
+
+
+            foreach (var task in tasks)
+            {
+                try
+                {
+                    task.WaitAndUnwrapException();
+                }
+                catch (Exception e)
+                {
+                    throw task.Exception ?? e;
```

Comment:
This can fail before the previous operations are completed, which can be pretty bad, no?

Reply:
Right, I switched to `Task.WaitAll` as you recommended